File: 1568-pseudo-palindromic-paths-in-a-binary-tree/pseudo-palindromic-paths-in-a-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func pseudoPalindromicPaths (root *TreeNode) int {
    return dfs(root, make(map[int]bool, 0))
}

func dfs(root *TreeNode, mem map[int]bool) int {
    if root == nil {
        return 0
    }
    
    mem[root.Val] = !mem[root.Val]
    if root.Left == nil && root.Right == nil {
        count := 0
        for _, v := range mem {
            if v {
                count += 1
            }
        }
        
        mem[root.Val] = !mem[root.Val]
        
        if count <= 1 {
            return 1
        }
        
        return 0
    }
    
    total := dfs(root.Left, mem) + dfs(root.Right, mem)
    mem[root.Val] = !mem[root.Val]
    
    return total
}

File: 1568-pseudo-palindromic-paths-in-a-binary-tree/pseudo-palindromic-paths-in-a-binary-tree.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
 *   var value: Int = _value
 *   var left: TreeNode = _left
 *   var right: TreeNode = _right
 * }
 */
object Solution {
  def pseudoPalindromicPaths(root: TreeNode): Int = {
    var bits = 0
    var count = 0
    def dfs(node: TreeNode): Unit = {
      if (node == null) return
      bits = bits ^ (1 << node.value)
      if (node.left == null && node.right == null) {
        if (isPseudoPalindromic(bits)) count += 1
      } else {
        dfs(node.left)
        dfs(node.right)
      }
      bits = bits ^ (1 << node.value)
    }
    dfs(root)
    count
  }

  private def isPseudoPalindromic(bits: Int): Boolean = {
    bits == 0 || (bits & (bits - 1)) == 0
  }
}